src/scope.ml

```diff
@@ -120,38 +120,93 @@ type mode =
       representation of the RHS. *)
 
+(** [get_implicitness t] gives the specified implicitness of the parameters of
+    a symbol having the (parser-level) type [t]. *)
+let get_implicitness : p_term -> bool list = fun t ->
+  let rec get_impl t =
+    match t.elt with
+    | P_Prod(xs,t) -> List.map (fun (_,_,impl) -> impl) xs @ get_impl t
+    | P_Impl(_,t)  -> false :: get_impl t
+    | P_Wrap(t)    -> get_impl t
+    | _            -> []
+  in
+  let impl = get_impl t in
+  (* Minimization of the data: the list does not end with [false]. *)
+  let rec rem_false l = match l with false::l -> rem_false l | _ -> l in
+  List.rev (rem_false (List.rev impl))
+
 (** [scope md ss env t] turns a parser-level term [t] into an actual term. The
     variables of the environment [env] may appear in [t], and the scoping mode
     [md] changes the behaviour related to certain constructors.  The signature
     state [ss] is used to hande module aliasing according to [find_qid]. *)
 let scope : mode -> sig_state -> env -> p_term -> tbox = fun md ss env t ->
-  (* Unique name generation for wildcards in LHS. *)
-  let fresh : unit -> string =
+  (* Unique pattern variable generation for wildcards in a LHS. *)
+  let fresh_patt : env -> tbox =
     let c = Pervasives.ref (-1) in
-    fun () -> Pervasives.incr c; Printf.sprintf "#%i" Pervasives.(!c)
+    fun env ->
+      let name = Pervasives.incr c; Printf.sprintf "#%i" Pervasives.(!c) in
+      let xs = List.map (fun (_,(x,_)) -> _Vari x) env in
+      _Patt None name (Array.of_list xs)
   in
+  (* Toplevel scoping function, with handling of implicit arguments. *)
   let rec scope : env -> p_term -> tbox = fun env t ->
-    let scope_domain : popt -> env -> p_term option -> tbox = fun pos env a ->
-      match (a, md) with
-      | (Some(a), M_LHS(_)) -> fatal a.pos "Annotation not allowed in a LHS."
-      | (None   , M_LHS(_)) ->
-          let xs = List.map (fun (_,(x,_)) -> _Vari x) env in
-          _Patt None (fresh ()) (Array.of_list xs)
-      | (None   , M_RHS(_)) -> fatal pos "Missing type annotation in a RHS."
-      | (Some(a), _       ) -> scope env a
-      | (None   , _       ) ->
-          (* We build a metavariable that may use the variables of [env]. *)
-          let vs = Env.vars_of_env env in
-          _Meta (fresh_meta (prod_of_env env _Type) (Array.length vs)) vs
+    (* Extract the spine. *)
+    let (h, args) = Syntax.get_args t in
+    (* Check whether application is marked as explicit in the head symbol. *)
+    let expl = match h.elt with P_Iden(_,b) -> b | _ -> false in
+    (* Scope the head and obtain the implicitness of arguments. *)
+    let h = scope_head env h in
+    let impl =
+      (* We avoid unboxing if [h] is not closed (and hence not a symbol). *)
+      if not (Bindlib.is_closed h) then [] else
+      match Bindlib.unbox h with
+      | Symb(s,_) when expl && s.sym_implicits = [] ->
+          fatal t.pos "Explicit application of symbol with no implicits."
```

Comment:
Why a fatal error here? It shouldn't matter to prefix any symbol by @.

Reply:
I think that we should at least give a warning here, since it really makes no sense to explicitly apply a function that has no implicit arguments. Of course this is not strictly necessary from the point of view of the implementation, but it might confuse someone reading the code (I was confused by an `@append` that I removed in the examples written by @FranckS).

Reply:
Ok, I prefer a warning.

Reply:
OK, I was pretty sure that Coq would reject such a thing, but apparently no. I'll just remove the check and make the code simpler.